### PR TITLE
test: add scaled timer integration test

### PR DESCRIPTION
**Pull request by akonradi**
Commit Message: Add an integration test for the HTTP connection idle scaled timeout
Additional Description:
Add an integration test that verifies that when the resource pressure
attached to the reduce_timeouts overload action changes, it affects
streams that were created previously.

This also sets up for creating integration tests for other scaled
timeouts.

Risk Level: low
Testing: ran new test
Docs Changes: n/a
Release Notes: n/a
Platform Specific Features: n/a

### CODE REVIEW

File: test/integration/overload_integration_test.cc [hunk 1]
```diff
@@ -213,16 +223,98 @@ TEST_P(OverloadIntegrationTest, StopAcceptingConnectionsWhenOverloaded) {
...
-  // request.
+  // Reduce load a little to allow the connection to be accepted.
   updateResource(0.9);
   test_server_->waitForGaugeEq("overload.envoy.overload_actions.stop_accepting_connections.active",
                                0);
+  EXPECT_TRUE(fake_upstreams_[0]->waitForHttpConnection(*dispatcher_, fake_upstream_connection_));
+  EXPECT_TRUE(fake_upstream_connection_->waitForNewStream(*dispatcher_, upstream_request_));
+  ASSERT_TRUE(upstream_request_->waitForHeadersComplete());
+  ASSERT_TRUE(upstream_request_->waitForData(*dispatcher_, 10));
+  upstream_request_->encodeHeaders(Http::TestResponseHeaderMapImpl{{":status", "202"}}, true);
```
Comment (antoniovicente) [inline test/integration/overload_integration_test.cc:234]: Seems like the meaning of this test changed, the request is now getting to the upstream.

File: test/integration/overload_integration_test.cc [hunk 2]
```diff
@@ -213,16 +223,98 @@ TEST_P(OverloadIntegrationTest, StopAcceptingConnectionsWhenOverloaded) {
...
+
+  // Wait for the proxy to notice and take action for the overload.
+  test_server_->waitForCounterGe("http.config_test.downstream_cx_idle_timeout", 1);
+  dispatcher_->run(Event::Dispatcher::RunType::NonBlock);
+
+  if (GetParam().downstream_protocol == Http::CodecClient::Type::HTTP1) {
+    // For HTTP1, Envoy will start draining but will wait to close the
+    // connection. If a new stream comes in, it will set the connection header
+    // to "close" on the response and close the connection after.
+    auto response = codec_client_->makeRequestWithBody(request_headers, 10);
```
Comment (antoniovicente) [inline test/integration/overload_integration_test.cc:308]: This seems strange.  downstream_cx_idle_timeout was incremented but the downstream connection is still open and accepting requests?

How long does Envoy keep the downstream connection in the draining state?

File: test/integration/overload_integration_test.cc (hunk 1 above)
Comment (akonradi) [reply test/integration/overload_integration_test.cc:234]: That was intentional. This test was checking two overload behaviors at once: first that connections would not be accepted, then that when they were, new requests would be rejected. This seemed like a good way to disentangle those two.

File: test/integration/overload_integration_test.cc (hunk 2 above)
Comment (akonradi) [reply test/integration/overload_integration_test.cc:308]: There's a configurable drain timeout, applied [here](https://github.com/envoyproxy/envoy/blob/ebdcf7cc28dfeb555a957096231fdec343de4709/source/common/http/conn_manager_impl.cc#L1185).

Comment (antoniovicente) [reply test/integration/overload_integration_test.cc:308]: Seems like drain timeout defaults to 5 seconds.  I'm not sure if it is possible to disable.  Should we consider scaling the drain timeout?

Comment (akonradi) [reply test/integration/overload_integration_test.cc:308]: It's so small that I'm not sure it's necessary.

Comment (antoniovicente) [reply test/integration/overload_integration_test.cc:308]: It's relative.  If we allow the downstream connection timeout to scale to 0 secs, 5 second drain timeout may seem like an eternity.

Comment (akonradi) [reply test/integration/overload_integration_test.cc:308]: Fair enough. It would be easy enough to add. Do you have a strong preference?

Comment (antoniovicente) [reply test/integration/overload_integration_test.cc:308]: Let's add a comment about it to the issue for scaled timeouts.  We can decide later what to do about it.

**Review by antoniovicente — APPROVED** [review]

**Review by ggreenway — APPROVED** [review]

**Comment by antoniovicente** [comment]
TSAN failure that looks related to this PR found while running tests for https://github.com/envoyproxy/envoy/pull/14282

https://dev.azure.com/cncf/4684fb3d-0389-4e0b-8251-221942316e06/_apis/build/builds/60163/logs/145

WARNING: ThreadSanitizer: data race (pid=15)
Write of size 1 at 0x7b5c00027290 by main thread:
 __tsan_mutex_destroy /home/brian/src/final/llvm-project/compiler-rt/lib/tsan/rtl/tsan_interface_ann.cpp:472:3 (overload_integration_test+0x2d6100a)
absl::Mutex::~Mutex() /proc/self/cwd/external/com_google_absl/absl/synchronization/mutex.cc:727:3 (overload_integration_test+0x853dd82)
Envoy::FakeStream::~FakeStream() /proc/self/cwd/./test/integration/fake_upstream.h:51:7 (overload_integration_test+0x3072cc2)
Envoy::FakeStream::~FakeStream() /proc/self/cwd/./test/integration/fake_upstream.h:51:7 (overload_integration_test+0x3074946)
Envoy::FakeStream::~FakeStream() /proc/self/cwd/./test/integration/fake_upstream.h:51:7 (overload_integration_test+0x307498f)
std::__1::default_delete<Envoy::FakeStream>::operator()(Envoy::FakeStream*) const /opt/llvm/bin/../include/c++/v1/memory:2363:5 (overload_integration_test+0x2f05036)
std::__1::unique_ptr<Envoy::FakeStream, std::__1::default_delete<Envoy::FakeStream> >::reset(Envoy::FakeStream*) /opt/llvm/bin/../include/c++/v1/memory:2618:7 (overload_integration_test+0x2db4930)
Envoy::OverloadScaledTimerIntegrationTest_CloseIdleHttpConnections_Test::TestBody() /proc/self/cwd/test/integration/overload_integration_test.cc:317:15 (overload_integration_test+0x2daf2c0)

**Comment by akonradi** [comment]
Sorry, that should be fixed by #14327

**Pull request merged**